File: python/Sorts/mergesort.py
def mergesort(array):
    #Base case
    if len(array) <= 1:
        return array
    else:
        #Divide the array in half
        mid = round(len(array) / 2)
        left = array[0:mid]
        right = array[mid:]

        #Sort each half of the array
        left = mergesort(left)
        right = mergesort(right)

        #Combine sorted halves
        return merge(left, right)

#Combine two sorted arrays, into one, keeping all the elements in sorted order
def merge(left, right):
    array = list()
    while len(left) > 0 and len(right) > 0:
        #Look at the first element of the left and right arrays
        #And add the smaller one to the output array
        if left[0] < right[0]:
            array.append(left[0])
            del left[0]
        else:
            array.append(right[0])
            del right[0]
    #When the loop ends either left or right will have element(s) left in it
    if len(left) > 0:
        return array + left
    if len(right) > 0:
        return array + right

File: c/algorithm/merge sort.c
void merge(int a[],int l,int m,int u)
{
	int i,j,n1,n2,k;
	
    n1=m-l+1;
	n2=u-m;
	
    int a1[n1],a2[n2];

	for(i=0;i<n1;i++)
		a1[i]=a[i+l];

	for(i=0;i<n2;i++)
		a2[i]=a[n1+i];

	i=j=k=0;

	while (i < n1 && j < n2)
    {
        if (a1[i] <= a2[j]) a[k] = a1[i++];
        else a[k] = a2[j++];
       
        k++;
    }

    while (i < n1) a[k++] = a1[i++];

    while (j < n2) a[k++] = a2[j++];

}

void merge_sort(int a[],int l,int u)
{
	int m;
	if(l<u)
	{
		m = (u+l)/2;

		merge_sort(a,l,m);
		merge_sort(a,(m+1),u);
        merge(a,l,m,u);
		
	}

}
